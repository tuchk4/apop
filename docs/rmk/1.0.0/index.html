<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Home - Documentation</title>

    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="actions.html">actions</a><ul class='methods'><li data-type='method'><a href="actions.html#.clear">clear</a></li><li data-type='method'><a href="actions.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="actions.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="shortcuts.html">shortcuts</a><ul class='methods'><li data-type='method'><a href="shortcuts.html#.clear">clear</a></li><li data-type='method'><a href="shortcuts.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="shortcuts.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="utils.html">utils</a><ul class='methods'><li data-type='method'><a href="utils.html#.parseArray">parseArray</a></li><li data-type='method'><a href="utils.html#.parseObject">parseObject</a></li><li data-type='method'><a href="utils.html#.reduce">reduce</a></li></ul></li></ul>
</nav>

<div id="main">
    

    



    









    


    <section class="readme">
        <article><h1>RMK</h1><p>Object transformation without mutations.</p>
<pre class="prettyprint source"><code>REMAKE = Function(data): newData
Formula = Function(REMAKE_1 ... REMAKE_n): Function(data): newData</code></pre><p>Pseudo example:</p>
<pre class="prettyprint source lang-js"><code>import rmk from 'rmk';
import toCamelCase from '@rmk/to-camel-case';
import clear from '@rmk/clear';
import update from '@rmk/update';

const formula = rmk(
  toCamelCase(),
  clear(),
  update({
    id: data => data.id + 10
  })
);

const object = [{
  id: 1,
  groupId: 1,
  name: null
}, {
  id: 2,
  groupId: 2,
  name: ''
}];

const newObject = rmk(object);
expect(newObject).toEqual([{
  id: 10,
  group_id: 1
}, {
  id: 20,
  group_id: 2
}])</code></pre></article>
    </section>






</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.4.3</a> on Mon Feb 20 2017 17:16:35 GMT+0200 (EET) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>
</body>
</html>