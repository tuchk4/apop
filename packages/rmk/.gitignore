.DS_Store
/.git
/.idea
/npm-debug.log
/node_modules
/dist
recursive
*.json
/*.js
/*.map
!__tests__
!src
!rollup-build.js
