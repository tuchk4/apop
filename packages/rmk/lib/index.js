import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import flow from 'lodash/flow';

const parseObject = (origin, actions) => {
  return flow(actions)(origin);
};

const parseArray = (origin, actions) => {
  return origin.map(element => parseObject(element, actions));
};

const reduce = (origin, actions) => {
  if (isArray(origin)) {
    return parseArray(origin, actions);
  } else if (isObject(origin)) {
    return parseObject(origin, actions);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...actions) => {
  return origin => reduce(origin, actions);
};
module.exports = rmk;
