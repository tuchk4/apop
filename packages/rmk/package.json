{
  "name": "rmk",
  "version": "2.4.0",
  "description": "Flow transformation data.",
  "main": "index.js",
  "scripts": {
    "prepublish": "yarn build",
    "build": "babel-node ./rollup-build.js",
    "jsdoc2md": "npm run jsdoc2md:actions",
    "jsdoc2md:actions": "jsdoc2md src/actions/*.js  src/actions/**/*.js > docs/ACTIONS.md",
    "jsdoc2md:rmk": "jsdoc2md src/index.js > docs/README.md"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "author": "marvelsrp <marvelsrp@gmail.com>",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/tuchk4/rmk/issues"
  },
  "jest": {
    "rootDir": "./packages"
  },
  "homepage": "https://github.com/tuchk4/rmk",
  "dependencies": {
    "jsdoc": "^3.4.3",
    "jsdoc-to-markdown": "^3.0.0"
  }
}