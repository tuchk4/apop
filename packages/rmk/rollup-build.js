import { rollup } from 'rollup';

import resolve from 'rollup-plugin-node-resolve';
import uglify from 'rollup-plugin-uglify';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import { minify } from 'uglify-es';
import babelrc from 'babelrc-rollup';

const babelConfig = {
  presets: [['es2015', { loose: true, modules: false }, 'stage-0']],
  plugins: ['external-helpers', 'transform-object-rest-spread'],
};

var rollUpConfig = {
  format: 'umd',
  sourceMap: 'inline',
  plugins: [
    resolve({
      jsnext: true,
      main: true,
      browser: true,
    }),
    commonjs(),
    babel(
      babelrc({
        addExternalHelpersPlugin: false,
        config: babelConfig,
        exclude: 'node_modules/**',
      })
    ),
  ],
};

async function make(config) {
  config.entry = config.entry || `./src/${config.moduleName}.js`;
  config.dest = config.dest || `./${config.moduleName}.js`;
  config.destMin = config.destMin || `./${config.moduleName}.min.js`;

  let bundle = await rollup({
    entry: config.entry,
    plugins: config.plugins,
  });



  await bundle.write({
    format: config.format,
    moduleName: config.moduleName,
    dest: config.dest,
    sourceMap: config.sourceMap,
  });
  await bundle.write({
    format: config.format,
    moduleName: config.moduleName,
    dest: config.destMin,
    sourceMap: config.sourceMap,
    plugins: [...rollUpConfig.plugins, uglify({}, minify)],
  });
  console.log(`make ${config.dest} and ${config.destMin}`);
}

(async function() {

  await make({...rollUpConfig,
    moduleName: 'rmk',
    entry: './src/rmk.js',
    dest: `./index.js`,
    destMin: `./index.min.js`
  });
  await make({...rollUpConfig,
    moduleName: 'rmkRecursive',
    entry: './src/recursive/rmk.js',
    dest: `./recursive/index.js`,
    destMin: `./recursive/index.min.js`
  });

  const actionBundles = [
    'clear', 'each', 'toCamelCase', 'toSnakeCase', 'rename', 'update',
    'recursive/clear', 'recursive/each',
    'recursive/toCamelCase', 'recursive/toSnakeCase',
    'recursive/rename', 'recursive/update'
  ];

  actionBundles.forEach(async moduleName => await make({...rollUpConfig, moduleName}));

})();
