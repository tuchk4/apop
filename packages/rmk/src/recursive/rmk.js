import rmkActions from './../actions';
import rmkParse from './../utils/parse';

const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default recursive;
