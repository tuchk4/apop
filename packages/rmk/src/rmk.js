import rmkActions from './actions';
import rmkParse from './utils/parse';

const singleActions = {};
const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  singleActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: false,
    });

  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const rmk = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: false }),
  singleActions
);
const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default Object.assign(rmk, { recursive });
