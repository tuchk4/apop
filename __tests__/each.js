import rmk from '../src/rmk';

const before = Object.seal({
  a: 1,
  b: 2,
  c: 3,
  d: 4,
});
const beforeDeep = Object.seal({
  a: 1,
  b: 2,
  c: 3,
  d: {
    a: 1,
    b: 2,
    c: 3,
    d: {
      a: 1,
      b: 2,
      c: 3,
      d: 4,
    },
  },
});
test('Each transform flow', () => {
  const formula = rmk(
    rmk.each((key, value) => {
      return { key: key + '_', value: value * 2 };
    })
  );

  const result = formula(before);
  expect(result).toMatchSnapshot();
});
test('Each transform shortcut', () => {
  const formula = rmk.each((key, value) => {
    return { key: key + '_', value: value * 2 };
  });

  const result = formula(before);
  expect(result).toMatchSnapshot();
});

test('Recursive flow each transform', () => {
  const formula = rmk.recursive(
    rmk.each((key, value) => {
      if (typeof value == 'number') {
        return { key: key + '_', value: value * 2 };
      } else {
        return { key: key + '_', value: value };
      }
    })
  );

  const result = formula(beforeDeep);
  expect(result).toMatchSnapshot();
});

test('Recursive shortcut each transform', () => {
  const formula = rmk.recursive.each((key, value) => {
    if (typeof value == 'number') {
      return { key: key + '_', value: value * 2 };
    } else {
      return { key: key + '_', value: value };
    }
  });

  const result = formula(beforeDeep);
  expect(result).toMatchSnapshot();
});
