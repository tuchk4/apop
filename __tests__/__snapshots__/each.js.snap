// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Each transform flow 1`] = `
Object {
  "a_": 2,
  "b_": 4,
  "c_": 6,
  "d_": 8,
}
`;

exports[`Each transform shortcut 1`] = `
Object {
  "a_": 2,
  "b_": 4,
  "c_": 6,
  "d_": 8,
}
`;

exports[`Recursive flow each transform 1`] = `
Object {
  "a_": 2,
  "b_": 4,
  "c_": 6,
  "d_": Object {
    "a_": 2,
    "b_": 4,
    "c_": 6,
    "d_": Object {
      "a_": 2,
      "b_": 4,
      "c_": 6,
      "d_": 8,
    },
  },
}
`;

exports[`Recursive shortcut each transform 1`] = `
Object {
  "a_": 2,
  "b_": 4,
  "c_": 6,
  "d_": Object {
    "a_": 2,
    "b_": 4,
    "c_": 6,
    "d_": Object {
      "a_": 2,
      "b_": 4,
      "c_": 6,
      "d_": 8,
    },
  },
}
`;
